Port existing zero_dim_dispatch optimizations from codegen and remove codegen capability.

gchanan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#37615 Port existing zero_dim_dispatch optimizations from codegen and remove codegen capability.**
* #37613 Move addmm broadcasting code from codegen layer to native layer.
* #37603 Move addbmm broadcasting from the codegen layer to native layer.
* #37582 Specify _th_ ops in CUDAUnaryOps macros so they are easier to find.

We probably missed a lot of these when we ported things from TH, but it's also probably not a huge deal.  There is only one left with fmod.

Differential Revision: [D21338030](https://our.internmc.facebook.com/intern/diff/D21338030)

gchanan committed:
Port existing zero_dim_dispatch optimizations from codegen and remove codegen capability.

We probably missed a lot of these when we ported things from TH, but it's also probably not a huge deal.  There is only one left with fmod.

[ghstack-poisoned]

gchanan committed:
Update on "Port existing zero_dim_dispatch optimizations from codegen and remove codegen capability."


We probably missed a lot of these when we ported things from TH, but it's also probably not a huge deal.  There is only one left with fmod.

[ghstack-poisoned]

gchanan committed:
Update on "Port existing zero_dim_dispatch optimizations from codegen and remove codegen capability."


We probably missed a lot of these when we ported things from TH, but it's also probably not a huge deal.  There is only one left with fmod.

Differential Revision: [D21338030](https://our.internmc.facebook.com/intern/diff/D21338030)

[ghstack-poisoned]

gchanan committed:
Update on "Port existing zero_dim_dispatch optimizations from codegen and remove codegen capability."


We probably missed a lot of these when we ported things from TH, but it's also probably not a huge deal.  There is only one left with fmod.

Differential Revision: [D21338030](https://our.internmc.facebook.com/intern/diff/D21338030)

[ghstack-poisoned]

bhosmer reviewed (approved):
LG, couple notes:
* haven't double checked that the new routing logic added in LegacyDefinitions.cpp replaces 100% of the pairs discovered by `discover_zero_dim_tensor_operations`
* happened to notice #37580, linked from #33094, which does the same codegen removal but without replacing it in the implementations. Looks like it's being worked on now, so might want to give them a heads up?

gchanan committed:
Update on "Port existing zero_dim_dispatch optimizations from codegen and remove codegen capability."


We probably missed a lot of these when we ported things from TH, but it's also probably not a huge deal.  There is only one left with fmod.

Differential Revision: [D21338030](https://our.internmc.facebook.com/intern/diff/D21338030)

[ghstack-poisoned]

gchanan committed:
Update on "Port existing zero_dim_dispatch optimizations from codegen and remove codegen capability."


We probably missed a lot of these when we ported things from TH, but it's also probably not a huge deal.  There is only one left with fmod.

Differential Revision: [D21338030](https://our.internmc.facebook.com/intern/diff/D21338030)

[ghstack-poisoned]

gchanan commented:
> haven't double checked that the new routing logic added in LegacyDefinitions.cpp replaces 100% of the pairs discovered by discover_zero_dim_tensor_operations

I checked the codegen using tool/git_add_generated_dirs.sh.  That's not particularly interesting though, we missed a lot of these optimizations when porting, these are just the ones we are seeing now.

gchanan commented:
> happened to notice #37580, linked from #33094, which does the same codegen removal but without replacing it in the implementations. Looks like it's being worked on now, so might want to give them a heads up?

I didn't look into that much detail but I think the analysis there is just wrong -- like, it's talking about lines in the codegen script and not looking at what is actually output in the codegen.  But again, I don't think it's a particularly big deal, I'll just make this PR apply the optimization.

gchanan committed:
Update on "Port existing zero_dim_dispatch optimizations from codegen and remove codegen capability."


We probably missed a lot of these when we ported things from TH, but it's also probably not a huge deal.  There is only one left with fmod.

Differential Revision: [D21338030](https://our.internmc.facebook.com/intern/diff/D21338030)

[ghstack-poisoned]

gchanan committed:
Update on "Port existing zero_dim_dispatch optimizations from codegen and remove codegen capability."


We probably missed a lot of these when we ported things from TH, but it's also probably not a huge deal.  There is only one left with fmod.

Differential Revision: [D21338030](https://our.internmc.facebook.com/intern/diff/D21338030)

[ghstack-poisoned]

gchanan committed:
Update on "Port existing zero_dim_dispatch optimizations from codegen and remove codegen capability."


We probably missed a lot of these when we ported things from TH, but it's also probably not a huge deal.  There is only one left with fmod.

Differential Revision: [D21338030](https://our.internmc.facebook.com/intern/diff/D21338030)

[ghstack-poisoned]
